### PR TITLE
feat(new codecs): introduce avro decoder

### PR DESCRIPTION
**Pull request by WarmSnowy**
This PR is base on [@cameronbraid](https://github.com/cameronbraid)'s PR #17521.

- rebase on origin/master and fix some build errors. [Comment](https://github.com/vectordotdev/vector/pull/17521#discussion_r1282028010)
- adjust Avro sorting in documentation 
- remove unused imports. [Comment](https://github.com/vectordotdev/vector/pull/17521#discussion_r1281990582)
- rename member of AvroDeserializerConfig. [Comment](https://github.com/vectordotdev/vector/pull/17521#discussion_r1281999967)
- use typedef to enhance readability. [Comment](https://github.com/vectordotdev/vector/pull/17521#discussion_r1282033667)
- convert magic number to const value. [Comment](https://github.com/vectordotdev/vector/pull/17521#discussion_r1282028810)
- make schema_definition support log_namespace.

### CODE REVIEW

**Review by drichards-87 — APPROVED** [review]

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:57]: 💬 Do you plan to change this?

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:124]: 💬  Do we want to accept empty `bytes` here? If yes, let's return `Ok` immediately. If not, return an error.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:119]: 🥜 Let's merge `parse_single` and `parse`.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:97]: 🥜 Tiny nit: Capitalize the first word in comments.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:207]: 💬 We should treat this as a `Vec<u8>>`. Let's also add UUID in a test.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:212]: 💬 Can we also add one simple roundtrip test? You can look at `lib/codecs/tests/native.rs` for inspiration.

**Review by pront — CHANGES_REQUESTED** [review]
Hi @WarmSnowy, thank you for this PR! 

Overall it is in a pretty good state. I left a few comments about implementation details.

**Comment by WarmSnowy** [comment]
Thanks for your review, I will check it tomorrow.

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:126]: AFAIK, [`null`](https://avro.apache.org/docs/1.11.1/specification/#primitive-types) is a AVRO type, and avro decoder schema will check it which I think is better than return `Ok(())`.

This is a demo which bad branch will fail.

```rust
use apache_avro::Schema;
use serde::{Deserialize, Serialize};


#[derive(Debug, Serialize,Deserialize,Clone)]
struct Test {
    a: i64,
    b: String,
}

#[derive(Debug, Serialize,Deserialize,Clone)]
struct Empty {
}

#[derive(Debug, Serialize,Deserialize,Clone)]
struct Test2 {
    a:i32,
    b:(),
}

fn main() {
    let raw_schema = r#"
    {
        "type": "null"
    }
    "#;
    let schema = Schema::parse_str(raw_schema).unwrap();

    let test = ();
    
    // Serialize data
    let value = apache_avro::to_value(test).unwrap();
    let value = value.resolve(&schema).unwrap();
    let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
    println!("serialize. data: {:?}, len: {}", bytes, bytes.len());

    // Deseriallize data
    let r = bytes.clone();
    let mut r = r.as_slice();
    let value = apache_avro::from_avro_datum(&schema, &mut r,None).unwrap();
    let value:() = apache_avro::from_value::<()>(&value).unwrap();
    println!("deseriallize. data: {:?}", value);

    // Bad deseriallize
    let raw_schema = r#"
    {
        "type": "int"
    }
    "#;
    let r = bytes.clone();
    let mut r = r.as_slice();
    let schema = Schema::parse_str(raw_schema).unwrap();
    let value = apache_avro::from_avro_datum(&schema, &mut r,None).unwrap();
    let value:() = apache_avro::from_value::<()>(&value).unwrap();
    println!("bad deseriallize. data:{:?}", value);
}
```

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:58]: No plans in the near future, probably after finishing the other task at my hand. Do I need to remove the `TODO` comment?

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs]: In the avro doc, [the uuid is a logic type annotates an string](https://avro.apache.org/docs/1.11.1/specification/#uuid), there maybe some trouble when avro encoding uuid if we treat uuid as `Vec<u8>`.

**Review by drichards-87 — APPROVED** [review]

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs]: 💬 Do we need the uuid hyphenated as well? This is the type of thing a roundtrip test would catch. See my suggestion bellow.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs:58]: Feel free to remove or edit the comment. I would like to understand better the intention here. Can you explain a bit more?

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs:126]: This makes sense, thank you! 

Let's add an early check:
```
if bytes.is_empty() {
    return Ok(smallvec![]);
}
```

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs]: I tried the following code which try to serialize the uuid data to avro format. If we use type `uuid` or `String` for uuid data, the encoding will success, but `Vec<u8>` will fail. So I think that let uuid data stay `String` format is better than treating uuid as Vec<u8>.

```rust
let raw_schema = r#"
{
    "type": "string",
    "logicalType":"uuid"
}
"#;
let schema = Schema::parse_str(raw_schema).unwrap();

// This branch success
let test = uuid!("67e55044-10b1-426f-9247-bb680e5fe0c8");
let test:String = test.as_hyphenated().to_string();
let value = apache_avro::to_value(test).unwrap();
let value = value.resolve(&schema).unwrap();
let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
println!("{:?}", bytes);

// This branch success
let test:Uuid = uuid!("67e55044-10b1-426f-9247-bb680e5fe0c8");
let value = apache_avro::to_value(test).unwrap();
let value = value.resolve(&schema).unwrap();
let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
println!("{:?}", bytes);

// This branch will faill
let test = uuid!("67e55044-10b1-426f-9247-bb680e5fe0c8");
let test:Vec<u8> = test.as_bytes().to_vec();
let value = apache_avro::to_value(test).unwrap();
let value = value.resolve(&schema).unwrap();
let bytes = apache_avro::to_avro_datum(&schema, value).unwrap();
println!("{:?}", bytes);
```

Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs]: This makes sense, thank you for double checking. We should include this in the tests. 
In general this PR looks good now! There are a few unresolved comments mainly about strengthening tests and then we can merge.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs:227]: I pushed a commit to fix timestamp conversions. 

Also, I noticed that `AvroValue::TimeMicros` and `AvroValue::TimeMillis` are converted to integers although the appropriate conversion would be `Duration`? As long as we can encode/decode this correctly it's fine to keep it as is. That further increases the need for a roundtrip test.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs]: Added a test case, please review and fix if you found any bugs.

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:231]: Here it is intended to convert AVRO schema definitions to vectors for checksumming by other modules similar to what syslog does. However others decoders like protobuf seem to be simple, so I removed the todo tag here by 
[9aeab12.](https://github.com/vectordotdev/vector/pull/18636/commits/9ff71550c4b9c087a56ccc01e3280aeadca889bc)

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs]: Here it is intended to convert AVRO schema definitions to vectors for checksumming by other modules similar to what syslog does. However others decoders like protobuf seem to be simple, so I removed the todo tag here by 
[9aeab12.](https://github.com/vectordotdev/vector/pull/18636/commits/9ff71550c4b9c087a56ccc01e3280aeadca889bc)

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:231]: Thanks for your help!
Letting `AvroValue::TimeMicros` and `AvroValue::TimeMillis` be `Duration` needs to introduce a new type for `Vrl` which may be a lots of work, I think that keeping it as original is better.

Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:231]: Do you have some suggest for roundtrip test? I looked `lib/codecs/tests/native.rs` and found that it use `quickcheck::Gen` to  generate test data and I didn't found how to generate the schema definition file. Could I do it just like `lib/codecs/tests/protobuf.rs` which use a simple roundtrip test case ?

Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs:231]: Hi @WarmSnowy, thank you for these updates!

You can look at this example [here](https://github.com/vectordotdev/vector/blob/master/lib/codecs/tests/native_json.rs#L10) instead. Let me know if there are any follow-up questions.

Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:231]: I submitted some commits to do roundtrip test.
Some avro type test items were omitted due to some detail differences like `i32`.

Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs:231]: You can add all those new strings to `allow.txt`.

File: lib/codecs/tests/data/avro/avro_generate_fixtures.patch
Comment (pront) [inline lib/codecs/tests/data/avro/avro_generate_fixtures.patch:1]: Thanks for adding such extensive tests!

Nit 🥜: Rename `lib/codecs/tests/data/avro/avro` to `lib/codecs/tests/data/avro/generated`.

We really want to move away from this patch practice. A better alternative is to add a new separate target under a `tests/bin` directory. Then we can execute it with a command like `cargo --bin generate-avro-fixtures` which overwrites an existing `lib/codecs/tests/data/avro/generated` directory. Finally, in the `README.md` we can just include this single command and a quick overview of what this generated folder contains and when it is required to regenerate the fixtures.

File: lib/codecs/tests/avro.rs
Comment (pront) [inline lib/codecs/tests/avro.rs:28]: 🥜 Since we are using `rstest`, can this boolean be a `#[case()]` and then merge `roundtrip_avro_fixtures2` and `roundtrip_avro_fixtures_reverse2`?

File: lib/codecs/tests/avro.rs
Comment (pront) [inline lib/codecs/tests/avro.rs:21]: Are these failing?

Comment (WarmSnowy) [reply lib/codecs/tests/avro.rs:21]: Yes, these failed because the difference of inner type of int which avro using `i32` but vrlvalue using `i64`. And the values of `map` may be serialized in different order.

File: .github/actions/spelling/patterns.txt
Comment (pront) [inline .github/actions/spelling/patterns.txt:225]: Is this needed now?

File: lib/codecs/tests/data/avro/avro_generate_fixtures.patch
Comment (WarmSnowy) [reply lib/codecs/tests/data/avro/avro_generate_fixtures.patch]: I deleted the patch and use feature to control it, `cargo test --features generate-avro-fixtures --package codecs --test avro -- generate_avro_fixtures` which overwrites an existing lib/codecs/tests/data/avro/generated test data.

File: lib/codecs/tests/avro.rs
Comment (pront) [reply lib/codecs/tests/avro.rs:23]: Could we relax/specialize the tests in order to cover these cases?

File: .github/actions/spelling/patterns.txt
Comment (WarmSnowy) [reply .github/actions/spelling/patterns.txt:225]: No need for that now, I will remove it.

File: lib/codecs/tests/data/avro/avro_generate_fixtures.patch
Comment (pront) [reply lib/codecs/tests/data/avro/avro_generate_fixtures.patch]: This is a very good step forward. 

However, I would still prefer this to be a separate bin target to fully decouple it from the rest of the Vector codebase. Also, semantically this is not a test so it might be confusing to future readers/maintainers.

To be more specific, in `Cargo.toml` add:
```
[[bin]]
name = "generate-avro-fixtures"
path = "lib/codecs/tests/bin/generate-avro-fixtures.rs"
```
And in `lib/codecs/tests/bin/generate-avro-fixtures.rs`:
```
fn main() {
    // run all generators
}
```

And finally invoke with:
```
cargo run --bin generate-avro-fixtures
```

File: lib/codecs/tests/avro.rs
Comment (WarmSnowy) [reply lib/codecs/tests/avro.rs:23]: I didn't find a easy method, avro uses `i32` for some types which will throw errors when meeting `i64` which used by VrlValue Integer.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:226]: Let's return an error here instead.

File: lib/codecs/tests/avro.rs
Comment (WarmSnowy) [reply lib/codecs/tests/avro.rs:23]: | test case        | error message                        | avro inner type       | vrl inner type |
| ---------------- | ------------------------------------ | --------------------- | -------------- |
| date             | Date expected, got Long              | Date(i32)             | i64            |
| decimal_var      | Decimal expected, got Array          | Decimal(Decimal)      | Vec<u8>        |
| duration         | Duration expected, got Array         | Duration(Duration)    | [u8; 12]       |
| fixed            | String expected for fixed, got Array | Fixed(usize, Vec<u8>) | Vec<u8>        |
| time_millis      | TimeMillis expected, got Long        | TimeMillis(i32)       | i64            |
| timestamp_micros | TimestampMicros expected, got String | TimestampMicros(i64)  | DateTime<Utc>  |
| timestamp_millis | TimestampMillis expected, got String | TimestampMillis(i64)  | DateTime<Utc>  |

Comment (WarmSnowy) [reply lib/codecs/tests/avro.rs:23]: For `map` type, could I convert the two items test case to one item test case to relax the test?

File: lib/codecs/tests/data/avro/avro_generate_fixtures.patch
Comment (WarmSnowy) [reply lib/codecs/tests/data/avro/avro_generate_fixtures.patch]: Is this [chore(avro codec): move fixtures generator to tests/bin](https://github.com/vectordotdev/vector/pull/18636/commits/dc63aeeb56781b567a0410596f275679fa33b5d0) ok ?

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:226]: Like TimestampMillis(i64) and TimestampMicros(i64), can we let them as is and let user deal it manually?

File: lib/codecs/tests/data/avro/avro_generate_fixtures.patch
Comment (pront) [reply lib/codecs/tests/data/avro/avro_generate_fixtures.patch]: It's great, thank you.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs:226]: Correct me if I am wrong, here we will panic when we hit this branch. I don't think we wan't to pursue more complicated solutions here. We should probably just do this:

```
return Err("format!("{} type is not supported", std::any::type_name::<T>())
```

File: lib/codecs/tests/avro.rs
Comment (pront) [reply lib/codecs/tests/avro.rs:23]: > For map type, could I convert the two items test case to one item test case to relax the test?

Yes, sounds good.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:196]: To workaround the roundtrip error you can do: 
`Ok(VrlValue::from(String::from_utf8_lossy(&bytes))),`

And then remove `fixed.avro` from the exclusion list.

File: lib/codecs/src/decoding/format/avro.rs
Comment (pront) [inline lib/codecs/src/decoding/format/avro.rs:186]: Let's also return an `Err` here for now.

In case you want to dig into this deeper:

An easier way to convert to bytes:
```
let decimal_bytes: Vec<u8> = d.try_into()?;
Ok(VrlValue::from(decimal_bytes))
```
            
The roundtrip for this fails at the encoder level, specifically when resolving the schema.  I guess problem has to do with specifying `"logicalType":"decimal"` in `decimal_var.avsc` but the bytes test passes.

File: lib/codecs/tests/avro.rs
Comment (pront) [reply lib/codecs/tests/avro.rs:23]: I also left a comment on how to fix `AvroValue::Fixed(_, bytes)`. 

For all the other types we can [do this](https://github.com/vectordotdev/vector/pull/18636#discussion_r1362513762). Workarounds should be doable but it's sufficient to return an error and unblock this PR.

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:196]: The [avro fixed](https://avro.apache.org/docs/1.11.1/specification/#fixed) is (usize, Vec<u8>) not String, converting it to String may have some side effect which `String::from_utf8_lossy` will replace some bytes.

[avro resolve_fixed](https://github.com/apache/avro/blob/bfc19adae5a55e119dc5ff3c1d9f8455e11557f3/lang/rust/avro/src/types.rs#L893-L905C4)

File: lib/codecs/src/decoding/format/avro.rs
Comment (WarmSnowy) [reply lib/codecs/src/decoding/format/avro.rs:226]: Or we can let it as it here?
```diff
diff --git a/lib/codecs/src/decoding/format/avro.rs b/lib/codecs/src/decoding/format/avro.rs
index 3bd6859792..fd0cf90c7e 100644
--- a/lib/codecs/src/decoding/format/avro.rs
+++ b/lib/codecs/src/decoding/format/avro.rs
@@ -211,19 +211,12 @@ pub fn try_from(value: AvroValue) -> vector_common::Result<VrlValue> {
         AvroValue::String(string) => Ok(VrlValue::from(string)),
         AvroValue::TimeMicros(time_micros) => Ok(VrlValue::from(time_micros)),
         AvroValue::TimeMillis(time_millis) => Ok(VrlValue::from(time_millis)),
-        AvroValue::TimestampMicros(micros) => Ok(VrlValue::Timestamp(
-            Utc.timestamp_micros(micros)
-                .single()
-                .ok_or("failed to convert TimestampMicros")?,
-        )),
-        AvroValue::TimestampMillis(millis) => Ok(VrlValue::Timestamp(
-            Utc.timestamp_millis_opt(millis)
-                .single()
-                .ok_or("failed to convert TimestampMillis")?,
-        )),
+        AvroValue::TimestampMicros(micros) => Ok(VrlValue::from(micros)),
+        AvroValue::TimestampMillis(millis) => Ok(VrlValue::from(millis)),
         AvroValue::Union(_, v) => try_from(*v),
         AvroValue::Uuid(uuid) => Ok(VrlValue::from(uuid.as_hyphenated().to_string())),
-        AvroValue::LocalTimestampMillis(_) | AvroValue::LocalTimestampMicros(_) => unimplemented!(),
+        AvroValue::LocalTimestampMillis(millis) => Ok(VrlValue::from(millis)),
+        AvroValue::LocalTimestampMicros(micros) => Ok(VrlValue::from(micros)),
     }
 }
```

Comment (pront) [reply lib/codecs/src/decoding/format/avro.rs:226]: > Like TimestampMillis(i64) and TimestampMicros(i64), can we let them as is and let user deal it manually?

Using the Utc API makes sense here. I was referring to all other types for which the mapping is not obvious/complex.

**Comment by pront** [comment]
Hi @WarmSnowy, this is really close to merging! Your work on this is greatly appreciated. Do you plan to work on this sometime soon? 

The only action item is to return `Err` for tricky to handle Avro types and documentation on which types are currently supported.

**Comment by Ion-manden** [comment]
@pront Pr. is up #19342

**Comment by pront** [comment]
Work was completed here: https://github.com/vectordotdev/vector/pull/19342

**Pull request closed without merging**
